libpromises/evalfunction.c

```diff
@@ -8450,6 +8451,191 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
+        thread_info->result = FnFailure();
+        return NULL;
+    }
+
+    ret = pthread_mutex_unlock(&(thread_info->mutex));
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to unlock mutex: %s",
+            GetErrorStrFromCode(ret));
+        thread_info->result = FnFailure();
+        return NULL;
+    }
+
+    thread_info->result = FnReturnContext(success);
+    return NULL;
+}
+
+static FnCallResult FnCallFileReadable(ARG_UNUSED EvalContext *const ctx,
+                                       ARG_UNUSED const Policy *const policy,
+                                       const FnCall *const fp,
+                                       const Rlist *finalargs)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+
+    struct ThreadInfo thread_info = {0};
+
+    if (finalargs == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires path as first argument",
+            fp->name);
+        return FnFailure();
+    }
+    thread_info.path = RlistScalarValue(finalargs);
+
+    long timeout = (finalargs->next == NULL) ? 0L
+                 : IntFromString(RlistScalarValue(finalargs->next));
+    assert(timeout >= 0);
+
+    if (timeout == 0)
+    {
+        char buf[1];
+        const int fd = open(thread_info.path, O_RDONLY | O_NONBLOCK);
+        if (fd < 0) {
+            Log(LOG_LEVEL_DEBUG, "Failed to open file '%s': %s",
+                thread_info.path, GetErrorStr());
+            return FnReturnContext(false);
+        }
+        else if (read(fd, buf, sizeof(buf)) < 0)
+        {
+            Log(LOG_LEVEL_DEBUG, "Failed to read from file '%s': %s",
+                thread_info.path, GetErrorStr());
+            close(fd);
+            return FnReturnContext(false);
+        }
+        close(fd);
+        return FnReturnContext(true);
+    }
+
+    struct timespec timespec = {0};
+    xclock_gettime(CLOCK_REALTIME, &timespec);
+    timespec.tv_sec += timeout;
+
+    int ret = pthread_mutex_init(&(thread_info.mutex), NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_mutex_lock(&(thread_info.mutex));
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to lock mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_create(&(thread_info.thread), NULL, &ThreadRoutine, &thread_info);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create thread: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_cond_init(&(thread_info.cond), NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize condition: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_cond_timedwait(&(thread_info.cond), &(thread_info.mutex),
+                                 &timespec);
```

Comment:
Can't we use `pthread_timedjoin_np()` or `pthread_tryjoin_np()` instead of the condition?

Reply:
> These functions are nonstandard GNU extensions; hence the suffix  "_np"
       (nonportable) in the names.

Meaning they will probably not work on exotics?

Reply:
Yes, but woth a try, I think.

Reply:
I don't think it's worth to spend more time on this